libmscore/realizedharmony.cpp

```diff
@@ -0,0 +1,555 @@
+///    gets the fraction duration for how long
+///    the harmony should be realized based
+///    on the HDuration set.
+///
+///    This is opposed to RealizedHarmony::duration()
+///    which returns the HDuration, which is the duration
+///    setting.
+///
+///    Specifying a parameter finds a duration based on the
+///    passed duration type while not specifying a parameter
+///    uses the setting set by the user for the specific
+///    harmony object.
+//--------------------------------------------------
+Fraction RealizedHarmony::getActualDuration(HDuration durationType) const
+      {
+      HDuration dur;
+      if (durationType != HDuration::INVALID)
+            dur = durationType;
+      else
+            dur = _duration;
+      switch (dur)
+            {
+            case HDuration::UNTIL_NEXT_CHORD_SYMBOL:
+                  return _harmony->ticksTilNext(false);
+                  break;
+            case HDuration::STOP_AT_MEASURE_END:
+                  return _harmony->ticksTilNext(true);
+                  break;
+            case HDuration::SEGMENT_DURATION:
+                  return toSegment(_harmony->parent())->ticks();
+                  break;
+            default:
+                  return Fraction(0, 1);
+            }
+      }
+
+//---------------------------------------------------
+//   getIntervals
+///   gets a map from intervals to TPCs based on
+///   a passed root tpc (this allows for us to
+///   keep pitches, but transpose notes on the score)
+///
+///   Weighting System:
+///   - Rank 0: 3rd, altered 5th, suspensions (characteristic notes)
+///   - Rank 1: 7th
+///   - Rank 2: 9ths
+///   - Rank 3: 13ths where applicable and (in minor chords) 11ths
+///   - Rank 3: Other alterations and additions
+///   - Rank 4: 5th and (in major/dominant chords) 11th
+//---------------------------------------------------
+QMap<int, int> RealizedHarmony::getIntervals(int rootTpc, bool literal) const
+      {
+      //RANKING SYSTEM
+      static const int RANK_MULT = 128;
+      static const int RANK_3RD = 0;
+      static const int RANK_7TH = 1;
+      static const int RANK_9TH = 2;
+      static const int RANK_ADD = 3;
+      static const int RANK_OMIT = 4;
+
+      static const int FIFTH = 7 + RANK_MULT*RANK_OMIT;
+
+      QMap<int, int> ret;
+
+      const ParsedChord* p = _harmony->parsedForm();
+      QString quality = p->quality();
+      int ext = p->extension().toInt();
+      const QStringList& modList = p->modifierList();
+
+      int omit = 0; //omit flags for which notes to omit (for notes that are altered
+                    //or specified to be omitted as a modification) so that they
+                    //are not later added
+      bool alt5 = false; //altered 5
+
+      //handle modifiers
+      for (QString s : modList) {
+            //find number, split up mods
+            bool modded = false;
+            for (int c = 0; c < s.length(); ++c) {
+                  if (s[c].isDigit()) {
```

Comment:
Another big TODO: a lot of this code is similar for MusicXML export, I wonder if we could somehow refactor some of it.

Reply:
I haven't looked at much of the MusicXML import/export, but if the code looks like what I think it does, it can probably be done. Especially after the enum stuff.